src/Storages/StorageFile.cpp

```diff
@@ -1621,17 +1636,37 @@ void ReadFromFile::createIterator(const ActionsDAG::Node * predicate)
         storage->distributed_processing);
 }
 
+void addPartitionColumnsToInfoHeader(Strings paths, ReadFromFormatInfo & info)
+{
+    for (const auto& path : paths)
+    {
+        auto map = VirtualColumnUtils::parsePartitionMapFromPath(path);
+        for (const auto& item : map)
+            info.source_header.insertUnique(ColumnWithTypeAndName(std::make_shared<DataTypeString>(), item.first));
+    }
+}
+
 void ReadFromFile::initializePipeline(QueryPipelineBuilder & pipeline, const BuildQueryPipelineSettings &)
 {
     createIterator(nullptr);
 
     size_t num_streams = max_num_streams;
 
     size_t files_to_read = 0;
+    Strings paths;
     if (storage->archive_info)
+    {
         files_to_read = storage->archive_info->paths_to_archives.size();
+        paths = storage->archive_info->paths_to_archives;
+    }
     else
+    {
         files_to_read = storage->paths.size();
+        paths = storage->paths;
+    }
+
+    if (getContext()->getSettingsRef().file_hive_partitioning)
+        addPartitionColumnsToInfoHeader(paths, info);
```

Comment:
Why do we need it here? As I see we add virtual columns to `info.source_header` but why?